Automattic/wp-calypso
# Jetpack Checkout: add purchase warning for JP Social overlapping plan in cart

Asana task: 1202411927350212-as-1203885046675411/f

## Proposed Changes

* In Calypso checkout for Jetpack products, there's a component that checks if the plan in the user's cart has features the owner already used  and displays a message if so.
* It was discovered Social did not display the message.
* There were two issues causing that, first the jetpack_complete plan does not return Jetpack Social (or Boost FWIW) as a plan feature.  Second, the logic in the cart warning component was hardcoded for each individual overlapping product, and those weren't added.
* This PR does two things:
* Refactor the component code so that rather than handpicking what features trigger the message, any match between the plan and the site will trigger the warning.
* Next because the complete product wasn't returning items that are products and not featured, I needed to get that data to come along with the feature data.  There is a get allFeaturesForPlan that returns the plan data, and Products is a viable part of the underlying object, but that function did not call features.
* Rather than just adding that to th existing function, I added a new function to main.ts that returns the Products only so it can be used in conjunction with the Features but not impact any other code.
* To be safe I added a Unit test for the new function so we'll be able to find out if something drops out of the product list.

![Screenshot 2023-03-14 at 12 40 02 PM](https://user-images.githubusercontent.com/12895386/225075866-de67985f-7c9c-4e15-9486-b668226f38e0.png)

## Testing Instructions

* You'll want a JN or other site that already has Social (that's what this PR addresses but it would also apply to boost) already.
* Add Jetpack Complete to your cart, and notice the lack of a warning
* Head on over to the Calypso Live Link (or roll your own with this PR) and append the slug /checkout/[your site here] and you should see the warning at the top of the cart
* Bonus round you can play with any other combination, like social and complete, akismet and security and so on.  Any combo should work and show the banner if there's overlap in the site products and plan.
* If you feel like running the unit test `yarn test-packages packages/calypso-products/test/get-all-products-for-plan.ts`

## Pre-merge Checklist
